src/System.IO.Abstractions.Extensions/IFileSystemExtensions.cs

```diff
@@ -11,5 +11,131 @@ public static IDirectoryInfo CurrentDirectory(this IFileSystem fileSystem)
+        /// <param name="directoryInfo">
+        /// The <see cref="IDirectoryInfo"/> for the directory that was created.
+        /// </param>
+        /// <returns>
+        /// An <see cref="IDisposable"/> to manage the directory's lifetime.
+        /// </returns>
+        public static IDisposable CreateDisposableDirectory(this IFileSystem fileSystem, out IDirectoryInfo directoryInfo)
+        {
+            return fileSystem.CreateDisposableDirectory(fileSystem.Path.GetRandomTempPath(), out directoryInfo);
+        }
+
+        /// <inheritdoc cref="CreateDisposableDirectory(IFileSystem, out IDirectoryInfo)"/>
+        /// <param name="path">
+        /// The full path to the directory to create.
+        /// </param>
+        /// <exception cref="ArgumentException">
+        /// If the directory already exists.
+        /// </exception>
+        public static IDisposable CreateDisposableDirectory(this IFileSystem fileSystem, string path, out IDirectoryInfo directoryInfo)
+        {
+            return fileSystem.CreateDisposableDirectory(path, dir => new DisposableDirectory(dir), out directoryInfo);
+        }
+
+        /// <inheritdoc cref="CreateDisposableDirectory(IFileSystem, string, out IDirectoryInfo)"/>
+        /// <param name="disposableFactory">
+        /// A <see cref="Func{T, TResult}"/> that acts as a factory method. Given the <see cref="IDirectoryInfo"/>, create the
+        /// <see cref="IDisposable"/> that will manage the its lifetime.
+        /// </param>
+        public static T CreateDisposableDirectory<T>(
+            this IFileSystem fileSystem,
+            string path,
+            Func<IDirectoryInfo, T> disposableFactory,
+            out IDirectoryInfo directoryInfo) where T : IDisposable
+        {
+            directoryInfo = fileSystem.DirectoryInfo.New(path);
+
+            if (directoryInfo.Exists)
+            {
+                throw CreateAlreadyExistsException(nameof(path), path);
+            }
+
+            directoryInfo.Create();
+
+            return disposableFactory(directoryInfo);
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="IFileInfo"/> using a random name from the temp path, and returns an <see cref="IDisposable"/>
+        /// that deletes the file when disposed.
+        /// </summary>
+        /// <param name="fileSystem">
+        /// The <see cref="IFileSystem"/> in use.
+        /// </param>
+        /// <param name="fileInfo">
+        /// The <see cref="IFileInfo"/> for the file that was created.
+        /// </param>
+        /// <returns>
+        /// An <see cref="IDisposable"/> to manage the file's lifetime.
+        /// </returns>
+        public static IDisposable CreateDisposableFile(this IFileSystem fileSystem, out IFileInfo fileInfo)
+        {
+            return fileSystem.CreateDisposableFile(fileSystem.Path.GetRandomTempPath(), out fileInfo);
+        }
+
+        /// <inheritdoc cref="CreateDisposableFile(IFileSystem, out IFileInfo)"/>
```

Comment:
this overload behaves a bit differently than the one it inherits the docs from. probably better to have a dedicated summary